bin/PurgeBadJobs.py

```diff
@@ -0,0 +1,127 @@
+#!/miniconda-latest/bin/python3
+# Script to purge jobs that have been queued for too long, or stuck in the created state for too long
+
+import logging
+import os
+import sys
+from configparser import ConfigParser
+from datetime import datetime, timedelta, timezone
+from time import sleep
+
+import pymongo
+from bson import ObjectId
+
+# Must be run in /kb/module
+sys.path.append("/kb/module")
+from lib.execution_engine2.utils.SlackUtils import SlackClient
+from lib.installed_clients.execution_engine2Client import execution_engine2
+from lib.execution_engine2.db.models.models import TerminatedCode, Status
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+config = ConfigParser()
+config.read(os.environ["KB_DEPLOYMENT_CONFIG"])
+ee2_endpoint = config.get(section="execution_engine2", option="ee2-url")
+slack_token = config.get(section="execution_engine2", option="slack-token")
+
+ee2 = execution_engine2(url=ee2_endpoint, token=os.environ["EE2_ADMIN_SERVICE_TOKEN"])
+slack_client = SlackClient(
+    slack_token, channel="#ee_notifications", debug=True, endpoint=ee2_endpoint
+)
+db_client = pymongo.MongoClient(
+    host=config.get(section="execution_engine2", option="mongo-host"),
+    port=int(config.get(section="execution_engine2", option="mongo-port")),
+    username=config.get(section="execution_engine2", option="mongo-user"),
+    password=config.get(section="execution_engine2", option="mongo-password"),
+    authSource=config.get(section="execution_engine2", option="mongo-database"),
+    authMechanism=config.get(section="execution_engine2", option="mongo-authmechanism"),
+    serverSelectionTimeoutMS=1000,
+)
+ee2_db = db_client.get_database(
+    config.get(section="execution_engine2", option="mongo-database")
+)
+ee2_jobs_collection = ee2_db.get_collection(
+    config.get(section="execution_engine2", option="mongo-jobs-collection")
+)
+
+
+def cancel(record):
+    job_id = str(record["_id"])
+    scheduler_id = record.get("scheduler_id")
+    cjp = {
+        "as_admin": True,
+        "job_id": job_id,
+        "terminated_code": TerminatedCode.terminated_by_automation.value,
+    }
+    print("About to cancel ee2 job", cjp)
+    ee2.cancel_job(params=cjp)
+    slack_client.cancel_job_message(
+        job_id=job_id,
+        scheduler_id=scheduler_id,
+        termination_code=TerminatedCode.terminated_by_automation.value,
```

Comment:
Would it be useful to know the source of the cancellation - e.g. is there more than one process that might be cancelling jobs?

Reply:
Possibly, but I don't know if anyone has even ever looked at these.

Reply:
Not sure what you mean by "these"

Reply:
The termination codes. We can make these more granular and granular for every reason, but not sure how useful it would be. Right now this Terminated is only used by the HeldJobsReaper or by someone deciding to include it when running a cancel_job call

Reply:
I wasn't thinking of more codes in the db, just maybe a new field in `slack_client.cancel_job_message` that contains `PurgeBadJobs.py` or something like that so that when you look at the slack channel you can tell which process is doing the cancelling. This can probably wait for another PR if so though.